File: directionReduction.py
def dirReduc(arr):
    dict = {"NORTH":"SOUTH","SOUTH":"NORTH","EAST":"WEST","WEST":"EAST"}
    res = []
    for i in arr:
        if res and dict[i] == res[-1]:
            res.pop()
        else:
            res.append(i)
    print(res)


dirReduc(["NORTH", "SOUTH", "SOUTH", "EAST", "WEST", "NORTH", "WEST"])


# def dirReduc(arr):
#     dir = " ".join(arr)
#     dir2 = dir.replace("NORTH SOUTH",'').replace("SOUTH NORTH",'').replace("EAST WEST",'').replace("WEST EAST",'')
#     dir3 = dir2.split()
#     return dirReduc(dir3) if len(dir3) < len(arr) else dir3

File: directionReduction.js
function dirRec(arr) {
    const opposites = { 'NORTH': 'SOUTH', 'EAST': 'WEST', 'SOUTH': 'NORTH', 'WEST': 'EAST' };
    const result = []
    arr.forEach(item => {
        if (result.length) {
            const prevItem = result.pop()
            if (prevItem !== opposites[item]) {
                result.push(prevItem)
                result.push(item)
            }
        } else {
            result.push(item)
        }
    })
    return result;
}

console.log(dirRec(["NORTH", "SOUTH", "SOUTH", "EAST", "WEST", "NORTH", "WEST"]));

/*
function dirReduc(arr){
    let str = arr.join("")
    let pattern = /NORTHSOUTH|EASTWEST|SOUTHNORTH|WESTEAST/g
    while(pattern.test(str)) {
        str = str.replace(pattern, '')
    }

    return str.match(/(NORTH|SOUTH|EAST|WEST)/g) || []
}
*/